cap-primitives/src/yanix/linux/fs/open.rs

```diff
@@ -0,0 +1,142 @@
+const RESOLVE_BENEATH: u64 = 0x08;
+
+#[repr(C)]
+#[derive(Debug, Default)]
+struct OpenHow {
+    oflag: u64,
+    mode: u64,
+    resolve: u64,
+}
+const SIZEOF_OPEN_HOW: usize = std::mem::size_of::<OpenHow>();
+
+/// Call the `openat2` system call.
+fn open_openat2(start: &fs::File, path: &Path, options: &OpenOptions) -> io::Result<fs::File> {
+    let oflags = compute_oflags(options);
+    let mode = options.ext.mode;
+
+    let path_cstr = CString::new(path.as_os_str().as_bytes())?;
+    let open_how = OpenHow {
+        oflag: u64::from(oflags.bits() as u32),
+        mode: u64::from(mode),
+        resolve: RESOLVE_BENEATH | RESOLVE_NO_MAGICLINKS,
+    };
+
+    // `openat2` fails with `EAGAIN` if a rename happens anywhere on the host
+    // while it's running, so use a loop to retry it a few times. But not too many
+    // times, because there's no limit on how often this can happen.
+    for _ in 0..4 {
+        unsafe {
+            match libc::syscall(
+                SYS_OPENAT2,
+                start.as_raw_fd(),
+                path_cstr.as_ptr(),
+                &open_how,
+                SIZEOF_OPEN_HOW,
+            ) {
+                -1 => match io::Error::last_os_error().raw_os_error().unwrap() {
+                    libc::EAGAIN => continue,
+                    libc::EXDEV => return escape_attempt(),
+                    errno => return other_error(errno),
+                },
+                ret => {
+                    let fd = ret as RawFd;
+
+                    #[cfg(debug_assertions)]
+                    {
+                        let check = open_manually_wrapper(
+                            start,
+                            path,
+                            options
+                                .clone()
+                                .create(false)
+                                .create_new(false)
+                                .truncate(false),
+                        )
+                        .expect("open_manually failed when open_openat2 succeeded");
+                        debug_assert!(
+                            is_same_file(start, &check)?,
+                            "open_manually should open the same inode as open_openat2"
+                        );
+                    }
+
+                    return Ok(fs::File::from_raw_fd(fd));
+                }
+            }
+        }
+    }
+
+    // Fall back to the manual-resolution path.
+    open_manually_wrapper(start, path, options)
+}
+
+lazy_static! {
+    static ref OPEN: fn(&fs::File, &Path, &OpenOptions) -> io::Result<fs::File> = {
+        // Test if `openat2` is supported. If so, we can use `open_openat2`.
+        // Otherwise, fall back to `open_manually`.
+        unsafe {
+            if let -1 = libc::syscall(
+                SYS_OPENAT2,
+                -1,
+                ptr::null::<libc::c_void>(),
+                ptr::null::<libc::c_void>(),
+                0,
+            ) {
```

Comment:
This looks like a perfect spot to reuse Alex's idea of having a global atomic bool set to `true` if the syscall doesn't exist (and falling back to an alternative), or `false` otherwise. This should be thread safe, plus make it a little bit more maintainable in the long run. This would then look something like:

```rust
static INVALID: AtomicBool = AtomicBool::new(false);
if !INVALID.load(SeqCst) {
    let rc = unsafe {
        libc::syscall(
            libc::SYS_openat2,
            start.as_raw_fd(),
            path_cstr.as_ptr(),
            &open_how,
            SIZEOF_OPEN_HOW,
       )
    };
    match rc {
        -1 => match io::Error::last_os_error().raw_os_error().unwrap() {
            libc::ENOSYS => INVALID.store(true, SeqCst);
        }
        // ...
    }
}

open_manually_wrapper(start, path, options)
```

FWIW though, I'm happy to follow up with this one in a subsequent PR.

Reply:
Cool, I think that'd be a good fit here. If you could add this in a follow up that'd be great!